### PR TITLE
CLN: clean-up sanitize_array series construction

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Some possible clean-up that I encountered when doing https://github.com/pandas-dev/pandas/pull/26848/

### CODE REVIEW

File: pandas/core/internals/construction.py [hunk 1]
```diff
@@ -549,59 +549,39 @@ def sanitize_array(data, index, dtype=None, copy=False,
...
-            if is_float_dtype(data.dtype) and is_integer_dtype(dtype):
-                try:
-                    subarr = _try_cast(data, True, dtype, copy,
-                                       True)
-                except ValueError:
-                    if copy:
-                        subarr = data.copy()
-            else:
-                subarr = _try_cast(data, True, dtype, copy, raise_cast_failure)
-        elif isinstance(data, Index):
```
Comment (jorisvandenbossche) [inline pandas/core/internals/construction.py]: It was already checked above that `data` is an ndarray, so can never be an index

File: pandas/core/internals/construction.py [hunk 2]
```diff
@@ -549,59 +549,39 @@ def sanitize_array(data, index, dtype=None, copy=False,
...
+                else:
+                    subarr = np.array(data, copy=False)
         else:
-
             # we will try to copy be-definition here
-            subarr = _try_cast(data, True, dtype, copy, raise_cast_failure)
+            subarr = _try_cast(data, dtype, copy, raise_cast_failure)
 
     elif isinstance(data, ExtensionArray):
-        if isinstance(data, ABCPandasArray):
```
Comment (jorisvandenbossche) [inline pandas/core/internals/construction.py]: The `extract_array(data, extract_numpy=True)` above already did this

File: pandas/core/internals/construction.py [hunk 3]
```diff
@@ -549,59 +549,39 @@ def sanitize_array(data, index, dtype=None, copy=False,
...
     data = extract_array(data, extract_numpy=True)
 
     # GH#846
     if isinstance(data, np.ndarray):
 
-        if dtype is not None:
-            subarr = np.array(data, copy=False)
-
+        if (dtype is not None
+                and is_float_dtype(data.dtype) and is_integer_dtype(dtype)):
```
Comment (jreback) [inline pandas/core/internals/construction.py]: future cleanup is to remove this *and* condition (as I *think* _try_cast already handles this)

File: pandas/core/internals/construction.py [hunk 4]
```diff
@@ -549,59 +549,39 @@ def sanitize_array(data, index, dtype=None, copy=False,
...
-
-        # everything else in this block must also handle ndarray's,
-        # becuase we've unwrapped PandasArray into an ndarray.
-
+        # it is already ensured above this is not a PandasArray
+        subarr = data
         if dtype is not None:
-            subarr = data.astype(dtype)
-
+            subarr = subarr.astype(dtype)
```
Comment (jreback) [inline pandas/core/internals/construction.py]: add copy=False here (not sure if it matters but can't hurt)

File: pandas/core/internals/construction.py [hunk 5]
```diff
@@ -677,10 +657,10 @@ def sanitize_array(data, index, dtype=None, copy=False,
     return subarr
 
 
-def _try_cast(arr, take_fast_path, dtype, copy, raise_cast_failure):
+def _try_cast(arr, dtype, copy, raise_cast_failure):
```
Comment (jreback) [inline pandas/core/internals/construction.py]: can you try to doc-string / type this?

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. small comments.

File: pandas/core/internals/construction.py [hunk 6]
```diff
@@ -677,10 +657,22 @@ def sanitize_array(data, index, dtype=None, copy=False,
     return subarr
 
 
-def _try_cast(arr, take_fast_path, dtype, copy, raise_cast_failure):
-
+def _try_cast(arr, dtype, copy, raise_cast_failure):
```
Comment (jreback) [inline pandas/core/internals/construction.py]: can you add types here?

**Review by jreback — APPROVED** [review]
lgtm, if you can add types to _try_cast would be great. merge on green.

File: pandas/core/internals/construction.py [hunk 7]
```diff
@@ -549,59 +549,39 @@ def sanitize_array(data, index, dtype=None, copy=False,
...
+        # it is already ensured above this is not a PandasArray
+        subarr = data
         if dtype is not None:
-            subarr = data.astype(dtype)
-
-        if copy:
-            subarr = data.copy()
+            subarr = subarr.astype(dtype, copy=copy)
+        elif copy:
+            subarr = subarr.copy()
```
Comment (jbrockmendel) [inline pandas/core/internals/construction.py]: Could we have EA here? If so, deep=True?

File: pandas/core/internals/construction.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/internals/construction.py]: Nice removing the fast path arg!

File: pandas/core/internals/construction.py [hunk 8]
```diff
@@ -677,10 +657,22 @@ def sanitize_array(data, index, dtype=None, copy=False,
...
 
-def _try_cast(arr, take_fast_path, dtype, copy, raise_cast_failure):
-
+def _try_cast(arr, dtype, copy, raise_cast_failure):
+    """
+    Convert input to numpy ndarray and optionally cast to a given dtype.
+
+    Parameters
+    ----------
+    arr : array-like
```
Comment (jbrockmendel) [inline pandas/core/internals/construction.py]: Any restriction on ndim?  Some of these functions are only called via the dataframe constructor.

File: pandas/core/internals/construction.py (hunk 6 above)
Comment (jreback) [reply pandas/core/internals/construction.py]: @jorisvandenbossche can you add types here?

**Comment by jreback** [comment]
@jbrockmendel ok with this?

**Comment by jbrockmendel** [comment]
yep

**Comment by jreback** [comment]
thanks @jorisvandenbossche

**Pull request merged**